runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/StreamingWorkerStatusReporter.java

```diff
@@ -0,0 +1,397 @@
+      // Sending a one time update. Use empty counter set for cumulativeCounters (2nd arg).
+      sendWorkerUpdatesToDataflowService(restartCounter, new CounterSet());
+    } catch (IOException e) {
+      LOG.warn("Failed to send harness startup counter", e);
+    }
+  }
+
+  /** Sends counter updates to Dataflow backend. */
+  private void sendWorkerUpdatesToDataflowService(
+      CounterSet deltaCounters, CounterSet cumulativeCounters) throws IOException {
+    // Throttle time is tracked by the windmillServer but is reported to DFE here.
+    streamingCounters.windmillQuotaThrottling().addValue(windmillQuotaThrottleTime.get());
+    if (memoryMonitor.isThrashing()) {
+      streamingCounters.memoryThrashing().addValue(1);
+    }
+
+    List<CounterUpdate> counterUpdates = new ArrayList<>(COUNTER_UPDATES_SIZE);
+
+    if (publishCounters) {
+      allStageInfo.get().forEach(s -> counterUpdates.addAll(s.extractCounterUpdates()));
+      counterUpdates.addAll(
+          cumulativeCounters.extractUpdates(false, DataflowCounterUpdateExtractor.INSTANCE));
+      counterUpdates.addAll(
+          deltaCounters.extractModifiedDeltaUpdates(DataflowCounterUpdateExtractor.INSTANCE));
+    }
+
+    // Handle duplicate counters from different stages. Store all the counters in a multimap and
+    // send the counters that appear multiple times in separate RPCs. Same logical counter could
+    // appear in multiple stages if a step runs in multiple stages (as with flatten-unzipped stages)
+    // especially if the counter definition does not set execution_step_name.
+    ListMultimap<Object, CounterUpdate> counterMultimap =
+        MultimapBuilder.hashKeys(counterUpdates.size()).linkedListValues().build();
+    boolean hasDuplicates = false;
+
+    for (CounterUpdate c : counterUpdates) {
+      Object key = getCounterUpdateKey(c);
+      if (counterMultimap.containsKey(key)) {
+        hasDuplicates = true;
+      }
+      counterMultimap.put(key, c);
+    }
+
+    if (hasDuplicates) {
+      extractUniqueCounters(counterUpdates, counterMultimap);
+    } else { // Common case: no duplicates. We can just send counterUpdates, empty the multimap.
+      counterMultimap.clear();
+    }
+
+    WorkItemStatus workItemStatus =
+        new WorkItemStatus()
+            .setWorkItemId(WINDMILL_COUNTER_UPDATE_WORK_ID)
+            .setErrors(failureTracker.drainPendingFailuresToReport())
+            .setCounterUpdates(counterUpdates);
+
+    dataflowServiceClient.reportWorkItemStatus(workItemStatus);
+
+    // Send any counters appearing more than once in subsequent RPCs:
+    while (!counterMultimap.isEmpty()) {
+      extractUniqueCounters(counterUpdates, counterMultimap);
+      dataflowServiceClient.reportWorkItemStatus(
+          new WorkItemStatus()
+              .setWorkItemId(WINDMILL_COUNTER_UPDATE_WORK_ID)
+              .setCounterUpdates(counterUpdates));
+    }
+  }
+
+  @VisibleForTesting
+  public void reportPeriodicWorkerMessage() {
```

Comment:
I don't see this used in any tests, can we make it private?

Reply:
done